src/render_window_async_display.cc

```diff
@@ -0,0 +1,44 @@
+#include "render_window.h"
+
+namespace node_sfml {
+namespace render_window {
+
+AsyncRenderWindowDisplay::AsyncRenderWindowDisplay(sf::RenderWindow* window,
+                                                   Nan::Callback* callback)
+    : Nan::AsyncWorker(callback) {
+  this->_window = window;
+}
+
+void AsyncRenderWindowDisplay::Execute() {
+  if (this->_window == nullptr) {
+    this->SetErrorMessage("No window provided!");
+    return;
+  }
+  if (!this->_window->isOpen()) {
+    this->SetErrorMessage("Closed window.");
+    return;
+  }
+
+  this->_window->setActive(true);
+  this->_window->display();
```

Comment:
I'm wondering whether it will have any thread problem that `_window->display()` runs in different threads.

Reply:
Since we're calling setActive(false) in the main thread before starting this one, I think it shouldn't be such an issue. SFML locks the resources so if we can't use the window it will report it in stderr.

On another project I'm doing the clear() -> draw() -> display() in a completely different thread but all the operation over shader in the current thread so far it seems to work fine as long as Nvidia users sets Threaded Optimization to "not default".

On my side I'd really prefer not having the main thread locked during the renderWindow#display() operations ^^ (So I'll be the consumer of this displayAsync and if I see any issues I'll try to solve them :) )

Reply:
`Execute()` is not in main thread I think.

Reply:
Yes, Execute() is on a different thread and that's totally expected (at least to me). The display function locks the current thread while waiting for FPS or for VSYNC.

Reply:
I mean `display()` run in another thread (in `Execute()`). And the main thread (JavaScript main event loop) continue running other logic without locking main thread.

E.g.

```js
setInterval(() => {
  // Draw a circle
  circle.draw();
});

while (true) {
    await window.displayAsync();  // display this `Execute()` is in another thread.
}
```

So `circle.draw()` may run in any time even if `window.dislayAsync()` -> `Execute()` -> `_window->display()` is current running.

I'm not what `displayAsync()` is mainly for. But what if we do `displayAsync()` just in the main event loop thread but in another tick? E.g.

```js
function displayAsync() {
  return new Promise((resolve, reject) => {
    process.nextTick(() => {
      window.display();
      resolve();
    });
  });
}
```

Reply:
So I think we can lock the main thread at the native side?

e.g. Set a critical section for drawable's `draw()`?

Reply:
I've taken a look at the critical sections I don't see a nice way to do it in native side.

The best I can see is using a library providing mutexes like https://www.npmjs.com/package/async-mutex

Reply:
Emmmm, after I taking a look at that code, I think it runs code at the same thread. Just like I said before:

```js
function displayAsync() {
  return new Promise((resolve, reject) => {
    process.nextTick(() => {
      window.display();
      resolve();
    });
  });
}
```

I mean set a CS at the cpp side of `draw` and `display`.

If it's a pure cpp project, how will we set mutex in vsync display?

Reply:
Generally in pure C++ project I never perform draw() operation in a different thread than the display() operation. I generally use 2 threads:
- main thread is doing the sprite.setPositon() sprite.setTexture() etc...
- draw thread is doing clear => draw => display

If I have to draw and display in two threads (like in this use case), that would be about having something like: `window.mutex` which holds a [sf::Mutex](https://www.sfml-dev.org/documentation/2.5.1/classsf_1_1Mutex.php) object and in each draw/display operation I would do `window.mutex.lock()` => `op` => `window.mutex.unlock()`.

It would be nice if there was a good threading solution like Ruby's [Ractor](https://github.com/ruby/ruby/blob/master/doc/ractor.md) so we could really have a part of the application that is dedicated to drawing. The JS Workers does not look very nice to use because you have to load up a different js file and I have no clue how it behave with stuff like pkg & Typescript :/

Reply:
So I mean lock a mutex in each draw in C++ side and window.display in C++ side.

Reply:
I think we should use [sf::Mutex](https://www.sfml-dev.org/documentation/2.5.1/classsf_1_1Mutex.php). I'll check when I have time to add a public readonly field to the Window class in C++ side so we can handle this mutex draw thing.